File: java/01-beginner/04-control-flow/SwitchCase.java
/**
 * LESSON: Switch Statement
 * An alternative to long if-else chains when comparing one variable.
 */
public class SwitchCase {
    public static void main(String[] args) {
        // ===== TRADITIONAL SWITCH =====
        int day = 3;
        System.out.println("Day number: " + day);
        switch (day) {
            case 1:
                System.out.println("Monday");
                break;  // break prevents "fall-through" to next case
            case 2:
                System.out.println("Tuesday");
                break;
            case 3:
                System.out.println("Wednesday");
                break;
            case 4:
                System.out.println("Thursday");
                break;
            case 5:
                System.out.println("Friday");
                break;
            case 6:
            case 7:
                System.out.println("Weekend!"); // cases 6 and 7 share same code
                break;
            default:
                System.out.println("Invalid day");
        }

        // ===== SWITCH WITH STRINGS =====
        String role = "admin";
        System.out.println("\nRole: " + role);
        switch (role.toLowerCase()) {
            case "admin":
                System.out.println("Full access granted");
                break;
            case "editor":
                System.out.println("Edit access granted");
                break;
            case "viewer":
                System.out.println("Read-only access");
                break;
            default:
                System.out.println("Unknown role");
        }

        // ===== ENHANCED SWITCH (Java 14+) =====
        // No break needed, uses arrow syntax
        String month = "MARCH";
        String season = switch (month) {
            case "DECEMBER", "JANUARY", "FEBRUARY" -> "Winter";
            case "MARCH", "APRIL", "MAY" -> "Spring";
            case "JUNE", "JULY", "AUGUST" -> "Summer";
            case "SEPTEMBER", "OCTOBER", "NOVEMBER" -> "Fall";
            default -> "Unknown";
        };
        System.out.println("\n" + month + " is in " + season);
    }
}

File: cpp-practice/01-beginner/04-control-flow/switch_case.cpp
/**
 * LESSON: Switch Statement
 */
#include <iostream>
#include <string>

int main() {
    // ===== TRADITIONAL SWITCH =====
    int day = 3;
    std::cout << "Day number: " << day << "\n";
    switch (day) {
        case 1: std::cout << "Monday\n"; break;
        case 2: std::cout << "Tuesday\n"; break;
        case 3: std::cout << "Wednesday\n"; break;
        case 4: std::cout << "Thursday\n"; break;
        case 5: std::cout << "Friday\n"; break;
        case 6:
        case 7:
            std::cout << "Weekend!\n"; break;
        default:
            std::cout << "Invalid day\n";
    }

    // ===== SWITCH WITH ENUM =====
    std::cout << "\n--- Switch with Enum ---\n";
    enum class Color { RED, GREEN, BLUE };
    Color c = Color::GREEN;

    switch (c) {
        case Color::RED:   std::cout << "Red\n"; break;
        case Color::GREEN: std::cout << "Green\n"; break;
        case Color::BLUE:  std::cout << "Blue\n"; break;
    }

    // ===== FALLTHROUGH (intentional) =====
    std::cout << "\n--- Fallthrough ---\n";
    int month = 2;
    switch (month) {
        case 1: case 3: case 5: case 7: case 8: case 10: case 12:
            std::cout << "31 days\n"; break;
        case 4: case 6: case 9: case 11:
            std::cout << "30 days\n"; break;
        case 2:
            std::cout << "28 or 29 days\n"; break;
    }

    // ===== SWITCH WITH INIT (C++17) =====
    std::cout << "\n--- Switch with Init (C++17) ---\n";
    switch (int val = 42 % 3; val) {
        case 0: std::cout << "Divisible by 3\n"; break;
        case 1: std::cout << "Remainder 1\n"; break;
        case 2: std::cout << "Remainder 2\n"; break;
    }

    return 0;
}
